src/log.cr

```diff
@@ -85,38 +85,41 @@
 #
 # ```
 # backend = Log::IOBackend.new
+# Log.builder.clear
 # Log.builder.bind "*", :warning, backend
 # Log.builder.bind "db.*", :debug, backend
 # Log.builder.bind "*", :error, ElasticSearchBackend.new("http://localhost:9200")
 # ```
 #
 # ### Configure logging from environment variables
 #
-# Include the following line to allow configuration from environment variables.
-#
-# ```
-# Log.setup_from_env
-# ```
-#
-# The environment variables `CRYSTAL_LOG_LEVEL` and `CRYSTAL_LOG_SOURCES` are used to indicate
```

Comment:
I'm not sure that this is an improvement. I think manually calling `Log.setup_from_env` makes sense because people may not want/expect it to be pulling something from the env. 

Is there a way to change the default log level without forcing people into `Log.setup_from_env`?

Or maybe I misunderstood and you still need to call `Log.setup_from_env`?

Reply:
If you want to change the default log level, you need to call again `setup_from_env`.

Using `Log.setup_from_env` enables out of the box minimal configuration via ENV var.

The app or framework is able to redefine the whole bindings and default level despite a built-in configuration. Even leaving out the flexibility of env vars.

Reply:
I guess my main concern is what percentage of people are going to be okay with these defaults.  It's probably fine for a development environment, but I have a feeling most people are going to just end up overriding the defaults anyway to setup things how they want; at which point were the defaults even worth it?  Given that `Log.setup_from_env` and `Log.builder.clear` are both one line, I'm still in favor of using the new defaults/config of `Log.setup_from_env` but requiring the user to call it if they want to use those defaults.

Reply:
I don't know, I find it pretty strange that if I write this:

```crystal
require "log"

Log.error { "hello world!" }
```

and run it, I get nothing as output. I would expect logging to go to STDOUT by default, without having to setup the logging. That should be the default.

Unrelated, but this example from the docs looks pretty strange:

```crystal
backend = Log::IOBackend.new
Log.builder.bind "*", :warning, backend
```

IOBackend? Which IO? Oh, I guess it uses STDOUT as a default? Not clear at all from that code. I think we should make the IOBackend constructor parameter required. Then you need to do `Log::IOBackend.new(STDOUT)`, which is longer but clearer in my opinion.

Reply:
We could actually have a subclass named `TTYBackend` or similar, that logs to `STDOUT` by default and uses coloring when the output is a tty

Reply:
@asterite Logging by default should be fine, I think everyone agrees on that. But `setup_from_env` does not only enable logging by default but also exposes configuration via environment variables and initializes the log builder so that you need to take that into account when configuring custom log setup.

I think the point is that for a simple program like your example logging should just work. But when you want to set up a custom logging configuration, you should not have to call `Log.builder.clear`. If you haven't setup anything explicitly that shouldn't be required.
This could be implemented using a flag which is set after initializing the logger and when a new log rule gets added and the flag is set, it clear the default rules. Afterwards the flag is disabled.

Reply:
@waj Coloring is fine but I don't think we need a another class for this.

Reply:
> but also exposes configuration via environment variables

In most Ruby libraries I've used, they always try to get a default config value from environment first. That's so nice because then you don't have to manually configure it. You know that whenever you need to use env vars, you can, and they have a precedence over everything else.

> This could be implemented using a flag which is set after initializing the logger and when a new log rule gets added and the flag is set, it clear the default rules.

I agree. Configuring the logger after the default has been set should reset it first.

Reply:
I don't think the configuration over environment variables is a problem by itself. But we need to figure out how to treat them when custom configuration comes into play. I don't think they can be easily integrated with a custom setup. And it their effect is reset like the default configuration that might be unexpected as well.

Reply:
> I don't think we need a another class for this

Sure, we don't. But it wouldn't hurt either and we could separate concerns, like checking if the fd is a tty or not is unnecessary for files. Log4j for example has a `ConsoleAppender`. I think it's much clear than creating a backend with `STDOUT`. Also it could switch to STDERR automatically or with configuration.